### PR TITLE
Add options to string type

### PR DESCRIPTION
**Pull request by zimt28**

### CODE REVIEW

**Review by zimt28 — COMMENTED** [review]
I'm wondering if this
https://github.com/ash-project/ash/blob/36749669caf8fe6f9f491f2635e05a26b96850e2/lib/ash/type/type.ex#L400-L405
should be changed to
```elixir
 def apply_constraints(type, term, constraints) do
   constraints = NimbleOptions.validate!(constraints, type.constraints())

   case type.apply_constraints(term, constraints) do 
     :ok -> {:ok, term} 
     other -> other 
   end 
 end 
```

File: lib/ash/type/string.ex [hunk 1]
```diff
@@ -34,6 +39,17 @@ defmodule Ash.Type.String do
   def apply_constraints(nil, _), do: :ok
 
   def apply_constraints(value, constraints) do
+    opts = NimbleOptions.validate!(constraints, @constraints)
```
[the remark this answers is not in the document]
Comment (zachdaniel) [reply lib/ash/type/string.ex:42]: Ah, yeah good point. That should probably be built into the `type` behavior...but this is fine for now.

**Comment by zachdaniel** [comment]
To fix the test failures, you'll want to add `constraints: [trim?: false]` to the `argument` here: https://github.com/ash-project/ash/blob/master/test/calculation_test.exs#L48

**Comment by zimt28** [comment]
@zachdaniel Done. Even though this _has_ to happen here, it's a little unexpected behavior ..

**Comment by zachdaniel** [comment]
You're correct, it is pretty strange... I'll have to think on it. I can't really see a way around it though.

**Comment by zimt28** [comment]
I've added a comment to the calculation docs. I'm wondering if values shouldn't actually be trimmed by default but just set to nil if empty.

**Comment by zachdaniel** [comment]
Ah, yeah, good point. `trim?` is kind of a different option from "empty means `nil`". Lets add another option here that implements that behavior. By combining the two, you could have `"  "` trimmed to `""` which would then be `nil`. Perhaps `allow_empty?: true/false`?

**Comment by zachdaniel** [comment]
Yeah, I think constraints should be applied generically.

**Comment by zachdaniel** [comment]
A credo error to fix, but for the other one, you'll have to update that argument to have `trim?: false`

File: lib/ash/type/string.ex [hunk 2]
```diff
@@ -34,39 +47,71 @@ defmodule Ash.Type.String do
...
+
+    trim_value? = opts[:trim?]
+    allow_empty? = opts[:allow_empty?]
+
+    trimmed_value = String.trim(value)
```
Comment (zachdaniel) [inline lib/ash/type/string.ex:55]: One thing to note here: trimming a large string could be expensive, and here we are trimming it at all times. But if `trim?` is `false` and `allow_empty` is `true`, we never actually use the `trimmed_value`. So we should choose a control flow pattern here that doesn't trim the string 100% of the time.

**Review by zachdaniel — APPROVED** [review]

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀

**Pull request merged**